# Robinhood API Tutorial

1. [Setting up your development environment](#Setting-up-your-development-environment)
2. [Logging in and getting started](#Logging-in-and-getting-started)
3. [Basic demo program](#Basic-Demo-Program)
4. [Exploring the API](#Exploring-the-API)
5. [2-Factor authentication](#2-Factor-Authentication)
5. [Getting data]()
6. [Implementing an automated trading strategy]()

## Setting up your development environment

We'll be using the [robin-stocks unofficial Robinhood API](https://github.com/jmfernandes/robin_stocksh) to build Python programs that interface with your Robinhood account. You'll need to be 18+ to open an account. This tutorial assumes you know some Python basics and already have Python3 set up. If you need a syntax refresher check out this [cheat sheet](https://www.codecademy.com/learn/learn-python-3/modules/learn-python3-syntax/cheatsheet) from Codecademy. There are plenty of tutorials available for setting up Python on your OS, so that will not be covered here. Keep in mind that standard market hours are 6:30AM - 1:00PM PST Mon-Fri.

First, we need to install the API:

In [1]:
pip install robin-stocks

Note: you may need to restart the kernel to use updated packages.


**Note**: your output will look different, as I'm running this in a [Jupyter Notebook](https://jupyter.org/index.html).

## Logging in and getting started

In [3]:
import robin_stocks as rh

user = "***REMOVED***"
pswd = "***REMOVED***"

rh.login(user, pswd)

ERROR: There was an issue loading pickle file. Authentication may be expired - logging in normally.
Enter Robinhood code for validation: 032085


{'access_token': 'eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1◼redacted◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼◼',
 'expires_in': 620681,
 'token_type': 'Bearer',
 'scope': 'internal',
 'refresh_token': '5DVHZSuBwV3EOybmXJwIXd10fHUnPs',
 'mfa_code': None,
 'backup_code': None,
 'detail': 'logged in with brand new authentication code.'}

Notice we were prompted to enter in a validation code. Robinhood sent me an SMS to authenticate the login before I can use the API. Then it generated an access token so that we can stay logged in without authenticating every time we acess the API in the next 419490 ms (7 min). We will delve into this process further later on. 

For now, we've sucessfully logged in and are ready to start using the API commands. Let's make sure everything is working by getting Apple's current pricing information:

In [50]:
rh.stocks.get_quotes("TSLA")

[{'ask_price': '1500.130000',
  'ask_size': 3,
  'bid_price': '1499.720000',
  'bid_size': 90,
  'last_trade_price': '1499.920000',
  'last_extended_hours_trade_price': None,
  'previous_close': '1476.490000',
  'adjusted_previous_close': '1476.490000',
  'previous_close_date': '2020-07-28',
  'symbol': 'TSLA',
  'trading_halted': False,
  'has_traded': True,
  'last_trade_price_source': 'nls',
  'updated_at': '2020-07-29T19:57:54Z',
  'instrument': 'https://api.robinhood.com/instruments/e39ed23a-7bd1-4587-b060-71988d9ef483/'}]

It looks like `last_trade_price` gives us the current stock price during trading hours. If we want to only access this information, we can specify with the `info` modifier: 

In [51]:
rh.stocks.get_quotes("TSLA", info="last_trade_price")

['1500.018300']

I opened Robinhood the app to double check that the output matches the actual current stock price. They matched, so we are good to go! 

Note: the info modifer works for all the commands that output a list like the one above.

## Basic Demo Program

We are ready to make a simple program that demonstrates some of the API's capabilities. 

Psuedo code:

```
login
output Tesla stock price
check if Tesla stock has fractional shares trading enabled
    if yes:
        buy $1 of $TSLA
        sell when it goes up by 1%
```

First I looked for a way check if a stock is fractionally tradeable on Robinhood. `stocks.get_quotes()` gave us all of the pricing info of a stock, but what about other non-pricing related information? BoOm - `stocks.get_instruments_by_symbols` comes in clutch by dumping out a list of everything else about a stock:

In [52]:
rh.stocks.get_instruments_by_symbols("TSLA")

[{'id': 'e39ed23a-7bd1-4587-b060-71988d9ef483',
  'url': 'https://api.robinhood.com/instruments/e39ed23a-7bd1-4587-b060-71988d9ef483/',
  'quote': 'https://api.robinhood.com/quotes/TSLA/',
  'fundamentals': 'https://api.robinhood.com/fundamentals/TSLA/',
  'splits': 'https://api.robinhood.com/instruments/e39ed23a-7bd1-4587-b060-71988d9ef483/splits/',
  'state': 'active',
  'market': 'https://api.robinhood.com/markets/XNAS/',
  'simple_name': 'Tesla',
  'name': 'Tesla, Inc. Common Stock',
  'tradeable': True,
  'tradability': 'tradable',
  'symbol': 'TSLA',
  'bloomberg_unique': 'EQ0000000003531703',
  'margin_initial_ratio': '0.5200',
  'maintenance_ratio': '0.4000',
  'country': 'US',
  'day_trade_ratio': '0.2500',
  'list_date': '2010-06-29',
  'min_tick_size': None,
  'type': 'stock',
  'tradable_chain_id': '0b29f27a-fbf1-4e11-b887-5597d422957f',
  'rhs_tradability': 'tradable',
  'fractional_tradability': 'tradable',
  'default_collar_fraction': '0.05'}]

The second-to-last item is `fractional_tradability`, which tells us if a stock is fractionally `tradable`. We can access this with the `info` modifier.

In [53]:
rh.stocks.get_instruments_by_symbols("TSLA", info="fractional_tradability")

['tradable']

Now we can log in, get the current stock price, and check if fractional trading is enabled. Next we need to use the `orders.order_buy_fractional_by_price` function found in the [robin_stocks documentation](https://robin-stocks.readthedocs.io/en/latest/functions.html#robin_stocks.orders.order_buy_fractional_by_price) so I can buy fractional shares in dollars:

![orders.order_buy_fractional_by_price explanation](img/fractional$.png)

this function takes in the stock symbol (TSLA in our case), how many dollars of the stock we want, and optionally, the time aspects of the order.

In [13]:
rh.orders.order_buy_fractional_by_price("TSLA", 1, 'gfd')

{'id': 'd3ac98b1-3df3-43bb-b98d-c2c28a5c4679',
 'ref_id': 'b0bbc2e0-5bf6-40c8-8a58-38374bad7ec0',
 'url': 'https://api.robinhood.com/orders/d3ac98b1-3df3-43bb-b98d-c2c28a5c4679/',
 'account': 'https://api.robinhood.com/accounts/◼◼◼◼redacted◼◼◼◼/',
 'position': 'https://api.robinhood.com/positions/◼◼◼◼redacted◼◼◼◼/e39ed23a-7bd1-4587-b060-71988d9ef483/',
 'cancel': 'https://api.robinhood.com/orders/d3ac98b1-3df3-43bb-b98d-c2c28a5c4679/cancel/',
 'instrument': 'https://api.robinhood.com/instruments/e39ed23a-7bd1-4587-b060-71988d9ef483/',
 'cumulative_quantity': '0.00000000',
 'average_price': None,
 'fees': '0.00',
 'state': 'unconfirmed',
 'type': 'market',
 'side': 'buy',
 'time_in_force': 'gfd',
 'trigger': 'immediate',
 'price': '1517.64000000',
 'stop_price': None,
 'quantity': '0.00065900',
 'reject_reason': None,
 'created_at': '2020-07-29T18:23:37.640616Z',
 'updated_at': '2020-07-29T18:23:37.640638Z',
 'last_transaction_at': '2020-07-29T18:23:37.640616Z',
 'executions': [],
 'ext

My order to buy $1, or 0.000659 shares at the time of writing, of Tesla stock executed. Next, we need a way of comparing the current price to the price we bought in at. We'll need to access the order details, as seen above. Luckily, `rh.orders.get_all_stock_orders()` will show us a list of all of the orders in our account and their IDs. I'm only going to access the exact order in this list because I'm not going to show you my entire account history, but if you make orders in between make sure you get the correct order's ID. 

In [55]:
rh.orders.get_all_stock_orders()[2]

{'id': 'd3ac98b1-3df3-43bb-b98d-c2c28a5c4679',
 'ref_id': 'b0bbc2e0-5bf6-40c8-8a58-38374bad7ec0',
 'url': 'https://api.robinhood.com/orders/d3ac98b1-3df3-43bb-b98d-c2c28a5c4679/',
 'account': 'https://api.robinhood.com/accounts/607459542/',
 'position': 'https://api.robinhood.com/positions/607459542/e39ed23a-7bd1-4587-b060-71988d9ef483/',
 'cancel': None,
 'instrument': 'https://api.robinhood.com/instruments/e39ed23a-7bd1-4587-b060-71988d9ef483/',
 'cumulative_quantity': '0.00065900',
 'average_price': '1517.45070000',
 'fees': '0.00',
 'state': 'filled',
 'type': 'market',
 'side': 'buy',
 'time_in_force': 'gfd',
 'trigger': 'immediate',
 'price': '1517.64000000',
 'stop_price': None,
 'quantity': '0.00065900',
 'reject_reason': None,
 'created_at': '2020-07-29T18:23:37.640616Z',
 'updated_at': '2020-07-29T18:23:38.883986Z',
 'last_transaction_at': '2020-07-29T18:23:37.753125Z',
 'executions': [{'price': '1517.58500000',
   'quantity': '0.00065900',
   'settlement_date': '2020-07-31',

Now that we have the order's ID, we can directly access the order info using `robin_stocks.orders.get_stock_order_info(orderID)`. 

In [27]:
rh.orders.get_stock_order_info('d3ac98b1-3df3-43bb-b98d-c2c28a5c4679')

{'id': 'd3ac98b1-3df3-43bb-b98d-c2c28a5c4679',
 'ref_id': 'b0bbc2e0-5bf6-40c8-8a58-38374bad7ec0',
 'url': 'https://api.robinhood.com/orders/d3ac98b1-3df3-43bb-b98d-c2c28a5c4679/',
 'account': 'https://api.robinhood.com/accounts/607459542/',
 'position': 'https://api.robinhood.com/positions/607459542/e39ed23a-7bd1-4587-b060-71988d9ef483/',
 'cancel': None,
 'instrument': 'https://api.robinhood.com/instruments/e39ed23a-7bd1-4587-b060-71988d9ef483/',
 'cumulative_quantity': '0.00065900',
 'average_price': '1517.45070000',
 'fees': '0.00',
 'state': 'filled',
 'type': 'market',
 'side': 'buy',
 'time_in_force': 'gfd',
 'trigger': 'immediate',
 'price': '1517.64000000',
 'stop_price': None,
 'quantity': '0.00065900',
 'reject_reason': None,
 'created_at': '2020-07-29T18:23:37.640616Z',
 'updated_at': '2020-07-29T18:23:38.883986Z',
 'last_transaction_at': '2020-07-29T18:23:37.753125Z',
 'executions': [{'price': '1517.58500000',
   'quantity': '0.00065900',
   'settlement_date': '2020-07-31',

Notice that `average_price` gives us the purchased price and `quantity` gives us the quantity of shares purchased.

Notice that the purchased price * purchased quantity = $1:

$ 1517.45 * 0.00065900 =  1  $

Thus if we want the current value of our dollar, we would do 

current stock price `float(rh.stocks.get_quotes("TSLA", info="last_trade_price"))` $*$ quantity purchased.

Float converts the price string into a float so we can use it in the multiplication.

Now that we can compare the current price to the purchased price, all we need is a way to sell the amount we purchased:

![orders.order_sell_fractional_by_quantity explanation](img/fractional-sell.png)

We must sell by quantity because the price will change over time. 

### Putting it all together:

**Note**: we must keep in mind that most of these commands allow us to input a string (1 stock) or a list of strings (multiple stocks). Because of this, all of them output a dictionary or list, even when we input 1 stock. Thus, if we want to access the output in a boolean operation, we must access the first element of the returned dict/list.

Let's look back at what we want our demo program to do:

```
login
output Tesla stock price
check if Tesla stock has fractional shares trading enabled
    if yes:
        buy $1 of $TSLA
        sell when $1 goes up by 1% ($1.01)
```

login:

In [34]:
# import robin_stocks as rh

# user = "test@gmail.com"
# pswd = "example"

# rh.login(user, pswd)

{'access_token': 'eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJleHAiOjE1OTY2NjcyMzgsInRva2VuIjoiUDB4SndjbTc2NWU0Q3lzUFQ4YUxmNTlYeUxya2pkIiwidXNlcl9pZCI6IjhhZjQ3YzcwLTdmMWYtNDQyZi1hNTEwLTBhODVhMmZkYzU2NiIsImRldmljZV9oYXNoIjoiYjJlYmFhNWMzZDE2MGIzNmFjNTM5ZTliMmIxZjAwZmQiLCJzY29wZSI6ImludGVybmFsIiwidXNlcl9vcmlnaW4iOiJVUyIsIm9wdGlvbnMiOnRydWUsImxldmVsMl9hY2Nlc3MiOmZhbHNlfQ.Nbx40yGl9NGY9ohubcZWenEE90Ig00aEqeRIUqA6IGGPAFqzVDpkkuGDkp02HNQSYCmGbhGw4va_zeoUe7ZM426ADA5bcXpvLCN2o_axH1O9Yslu46Po64sCsGO6V_dy7vZs4PZR3-fXClMHF7Qe4nFk7RdVaUz69p6p6ExV1OkX4ywD7Za5oFuNVIGTG2rgbtuO7uloI3qfWfz1e7b7__p8FnglJ0CMa6gtuGCqvSKD4ubK8g9HRNZAFXLJRAZe_PW2ehYfWjIedgsK0eOJtVTf4kiAyLOqkKscStKmI85odz1j01jhezmSJmocL1gHS7ACtvb4xPDfXhETwXzK5A',
 'token_type': 'Bearer',
 'expires_in': 86400,
 'scope': 'internal',
 'detail': 'logged in using authentication in robinhood.pickle',
 'backup_code': None,
 'refresh_token': '5DVHZSuBwV3EOybmXJwIXd10fHUnPs'}

output Tesla stock price:

In [82]:
print("Current Tesla stock price: " + rh.stocks.get_quotes("TSLA", info="last_trade_price")[0])

Current Tesla stock price: 1499.990000


check if Tesla stock has fractional shares trading enabled. If it does, buy a dollar of TSLA and sell when it goes up by 1% ($1.01):

In [79]:
if (rh.stocks.get_instruments_by_symbols("TSLA", info="fractional_tradability")[0] == 'tradable'):
    print("Fractional trading is enabled for TSLA")
    print("Attempting to buy $1 of TSLA")
    #print(rh.orders.order_buy_fractional_by_price("TSLA", 1, 'gfd')) # buy $1 of TSLA, order good for day
    z
    order_ID = rh.orders.get_all_stock_orders(info='id')[2] # get ID of latest order
    purchased_price = float(rh.orders.get_stock_order_info(order_ID)['average_price'])
    purchased_quantity = float(rh.orders.get_stock_order_info(order_ID)['quantity'])
    while True: # this is bad practice because we will make tons of API calls, but this is just a simple demo so it's okay
        current_dollar_value = float(rh.stocks.get_quotes("TSLA", info="last_trade_price")[0])*purchased_quantity
        if ( current_dollar_value >= 1.01 ):
            print("Attempting to sell "+current_dollar_value+" of TSLA")
            rh.orders.order_sell_fractional_by_quantity("TSLA", purchased_quantity, 'gfd')
else: 
    print("Tesla isn't fractionally tradeable")

Fractional trading is enabled for TSLA
Attempting to buy $1 of TSLA
Attempting to sell  0.9974933500000001  of TSLA


## Exploring the Documentation

The [API's documentation](https://robin-stocks.readthedocs.io/en/latest/) describes how to use the API and its various functions. Anything you can do in the Robinhood app, you can access with the robin-stocks API. To find out more about the available functions in the API, click [here](https://robin-stocks.readthedocs.io/en/latest/functions.html). Check out [robin-stocks on GitHub](https://github.com/jmfernandes/robin_stocks) to learn more about how the API works. 

## 2-Factor Authentication - work in progress

Notice how this will prompt 2-factor authentication.
This means the user has to input a login code sent by Robinhood via SMS/email every time the login function is executed. This would be very annoying if setting up an automated trading system. 
We can get around this by enabling 2 Factor Authentication and logging in with an authentication code. To do this, log in to Robinhood and go to settings, then scroll down and enable 2FA, then choose Authentication App, click "Can't Scan It?" and copy the code. (see GIF)

![settings](img/2FA.GIF)


In [5]:
import robin_stocks as rh

user = "test@gmail.com"
pswd = "test"
qr = "test"

rh.login(username = user, password = pswd, qr_code = qr)
rh.print_quote("AAPL")

this concludes the robin-stocks API tutorial. If you have any questions feel free to reach out to me (Ravi Riley).

# Building an Automated Trading Strategy

## Getting historical data

When implementing an automated trading strategy, you'll need a way to get reliable historical data for various stocks. I built a class to accomplish this built upon [Pandas Datareader](). 

## Implementing an Automated Trading Strategy

I'm not going to show you exactly how I made my own automated trading algorithm, but I will show you how to *implement* your own strategy using the Robinhood API. 